### ISSUE TITLE
Any plans to make new release?

### ISSUE THREAD

**Issue by kloczek** [libsdl-org/SDL_image#182]
I think that it would be good to flush currently committed changes and make new release :)

**Comment by illume** [comment]
Wondering about this too.

**Comment by madebr** [comment]
@1bsyl 
What are the plans on this matter?
What extra functionality do we want included?
Are there blockers?
What test coverage do we need?

**Comment by slouken** [comment]
I agree this would be good. SDL_image doesn't have any risky changes, the only blocker here is making sure M1 mac support works with the new framework.

**Comment by slouken** [comment]
Yep, we're making good strides towards a release candidate.

**Comment by slouken** [comment]
@madebr, @smcv, I think that except for CMake support, we're ready to cut a release-candidate build. I'd like to get CMake support in for the first RC, so it can get as much eyes on it as possible. What needs to be done at this point, before we do that?

**Comment by madebr** [comment]
I don't see any blockers, just some lack of test coverage + missing CMake review

Are we fine with the naming and behavior or the cmake configuration variables?
It will be hard to rename/change their meaning once they're out into the wild.
e.g. I would rename the `*_SHARED` options with `*_DYNAMIC` becauce `SHARED` in CMake-speak means shared libraries where in SDL_image we mean shared+dynamic.

The current cmake script also has a somewhat different philosophy then autotools.
Whereas most autotools scripts will look for dependencies and automatically disable some option when some system dependency is not found, CMake will complain hard and requires intervention from the user to either disable that feature or provide the dependency.
Having a packagers background myself, I like the latter for not providing any surprises and being certain a SDL_image build effectively has the options I passed the configuration options for.
But perhaps, you have a different view on this.


Missing test coverage
- We don't test MSVC on CI. From the lack of issues, I think it is working though, but it would be nice to have.
I think we'll need to get sdl2 from a binary tarball from the [releases tab](https://github.com/libsdl-org/SDL/releases).
The current minimum required SDL version is 2.0.8, so it would be nice to have these available on github too (to reduce network traffic between github.com and libsdl.org)
- We don't test whether the created `SDL2_imageConfig.cmake` is actually working.
I think testing this is important as this is the way *nix people will add SDL_image to their project.
- Testing vendoring SDL_image would also be useful I think, since we receive quite a few issues about this.

**Comment by slouken** [comment]
Yes, renaming the options where it makes sense is fine, and we can follow that same convention for the other SDL libraries.

Having CMake hard-fail with missing dependencies is fine, but in that case we should have AVIF, JXL, TIFF, and WebP off by default, to match the other build environments and provide a successful self-contained build. We can convert autotools to do that as well, @smcv, do you think that makes sense? (Although, Linux distributions should probably build with support for those formats enabled)

**Comment by slouken** [comment]
SDL 2.0.8 (and other release archives) and https://www.libsdl.org/release

**Comment by madebr** [comment]
I was mentioning it to avoid web traffic every time CI runs.

**Comment by sezero** [comment]
> Having CMake hard-fail with missing dependencies is fine, but
> [...] 
> We can convert autotools to do that as well,

No SDL library (SDL itself) behaved like this before. If we ever do that, we should document it clearly.

**Comment by slouken** [comment]
> I was mentioning it to avoid web traffic every time CI runs.

We have the 2.0.8 release tagged. Is that enough, or do we need to create a GitHub release and upload the archives?

**Comment by slouken** [comment]
> > Having CMake hard-fail with missing dependencies is fine, but
> > [...]
> > We can convert autotools to do that as well,
> 
> No SDL library (SDL itself) behaved like this before. If we ever do that, we should document it clearly.

Right. I think we can do it in the context of defaulting non-included codecs off, and defaulting included codecs on. That way someone can just run configure and build without having to download additional dependencies, and they'll get exactly what they expect.

Maybe at the end of configure we should output a list of codecs and ON/OFF state for them?

**Comment by sezero** [comment]
> Maybe at the end of configure we should output a list of codecs and ON/OFF state for them?

That's a good idea, for both autotools and cmake.

**Comment by madebr** [comment]
> We have the 2.0.8 release tagged. Is that enough, or do we need to create a GitHub release and upload the archives?

MSVC will need the binaries, or we would have to rebuild again from scratch (which takes time + adds another possible failure mode).
I don't think you can add files to tags, so you'll have to create a 2.0.8 release.


> Maybe at the end of configure we should output a list of codecs and ON/OFF state for them?

Great idea. I'll look at the SDL recipe for ideas.
There is also [`FeatureSummary`](https://cmake.org/cmake/help/latest/module/FeatureSummary.html). I haven't encountered it in the wild, but perhaps it's useful once mastered.

**Comment by slouken** [comment]
Okay, SDL 2.0.8 has been added to GitHub as a release

**Comment by smcv** [comment]
> hard-fail with missing dependencies

I like the tri-state behaviour seen in some Autotools projects, which Meson has copied for its options of type `feature`:

* `--disable-foo` (Autotools) or `-Dfoo=disabled` (Meson) unconditionally disables foo, even if found
* default behaviour is to enable foo if and only if the necessary library was found ("automagic dependency")
* `--enable-foo` or `-Dfoo=enabled` makes foo a hard requirement, failing with an error if it can't

It looks like the way to do that in CMake is to replace the `option` with a `CACHE STRING` variable with three choices, either `OFF`, `ON`, `FORCE_ON` or `OFF`, `AUTO`, `ON`: https://reviews.llvm.org/D40050, https://cmake.org/pipermail/cmake/2016-October/064342.html. I don't know whether one set of names for the three states is more conventional than any other.

This is particularly convenient with Meson's `-Dauto_features=(enabled|disabled)`, which sets all `feature` options that were not otherwise specified - useful for behaviours like "enable everything except AVIF" (which you'd spell as `-Dauto_features=disabled -Davif=enabled` if this was a Meson project). Maybe a mechanism similar to `SDL2IMAGE_DEPS_SHARED` and `SDL2IMAGE_VENDORED` could work for that?

**Comment by sezero** [comment]
> Okay, SDL 2.0.8 has been added to GitHub as a release

While you are at it, I suggest adding 2.0.9 (current minimum for SDL_mixer)
and 2.0.10 (current minimum for SDL_ttf). Then the only missing one would be
2.0.12.

**Comment by madebr** [comment]
Another important thing to consider is only the cmake build script will install a `SDL2_imageConfig.cmake` file.
Perhaps we will need to create `sdl2_image-config.cmake.autotools.in`, [similar to what SDL is doing](https://github.com/libsdl-org/SDL/blob/main/sdl2-config.cmake.in).

**Comment by slouken** [comment]
> Another important thing to consider is only the cmake build script will install a `SDL2_imageConfig.cmake` file. Perhaps we will need to create `sdl2_image-config.cmake.autotools.in`, [similar to what SDL is doing](https://github.com/libsdl-org/SDL/blob/main/sdl2-config.cmake.in).

Possibly. @smcv, do you think that's a good idea for autotools based builds for distributions? Or do you think distributions will start building with cmake by default?

**Comment by smcv** [comment]
> smcv, do you think that's a good idea for autotools based builds for distributions?

Yes, I think so. Recent SDL2 and dbus releases are good examples of an Autotools build that makes itself CMake-friendly.

> Or do you think distributions will start building with cmake by default?

I would expect most distributions to be cautious about adopting a new build system immediately. In Debian, I would certainly want to get the new SDL_image into `testing` using Autotools, and then consider switching to CMake afterwards, when I have a known-good Autotools build to compare it to.

Unix distributions tend to treat Autotools as "better the devil we know" - many important GNU and GNU-adjacent packages are Autotools-only, so we have to know how to deal with it *anyway*. The necessary workarounds for its disadvantages are well-known to packagers, and relatively consistent between different upstream packages.

**Comment by slouken** [comment]
> Another important thing to consider is only the cmake build script will install a `SDL2_imageConfig.cmake` file. Perhaps we will need to create `sdl2_image-config.cmake.autotools.in`, [similar to what SDL is doing](https://github.com/libsdl-org/SDL/blob/main/sdl2-config.cmake.in).

I created an issue to track this: https://github.com/libsdl-org/SDL_image/issues/276

@madebr, is there anything else that you'd like to see done before we create a release candidate for people to try out before release?

**Comment by madebr** [comment]
SDL_image's cmake support is in a nice state, it's only the autotools one that is stopping me for calling it good to go.
I'll post in #276 the bits I'm missing from autotools.

**Comment by slouken** [comment]
@madebr, we should hook up Visual Studio and Xcode support like your PR for SDL as well.

**Comment by madebr** [comment]
Another todo: in SDL https://github.com/libsdl-org/SDL/pull/5727, I'm currently adding a test for the cmake config file.
It's still a wip there because some platforms don't make live easy.
A similar test would also be useful here to have some certainty the cmake config files are actually working.

**Comment by slouken** [comment]
@madebr, just a sanity check, do you know of anything that needs to be done before final release?

**Comment by madebr** [comment]
I checked the VC sdk of the pre-release, and just tested the mingw pre-release.
Luckily I did, because the one of SDL_image and SDL2 contained a typo.

For the rest, I don't know what is missing.
I guess people will find ways to break it.

**Comment by smcv** [comment]
FYI, release candidate 2.5.2 went out in today's update for the Steam container runtime beta branch.

**Comment by peerless2012** [comment]
Hope it will publish soon, i can't wait to use cmake to build.
